lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,367 @@
+//===- MergeIfs.cpp -------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Arc/ArcOps.h"
+#include "circt/Dialect/Arc/ArcPasses.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-merge-ifs"
+
+namespace circt {
+namespace arc {
+#define GEN_PASS_DEF_MERGEIFSPASS
+#include "circt/Dialect/Arc/ArcPasses.h.inc"
+} // namespace arc
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+using namespace arc;
+
+namespace {
+struct MergeIfsPass : public arc::impl::MergeIfsPassBase<MergeIfsPass> {
+  void runOnOperation() override;
+  using MergeIfsPassBase::MergeIfsPassBase;
+  using MergeIfsPassBase::numIfsMerged;
+  using MergeIfsPassBase::numOpsMovedFromBetweenIfs;
+  using MergeIfsPassBase::numOpsMovedToUser;
+  using MergeIfsPassBase::numOpsSunk;
+};
+
+/// A helper to perform the op sinking within a specific block.
+struct Sinker {
+  MergeIfsPass &pass;
+  Block &rootBlock;
+  bool anyChanges;
+
+  Sinker(MergeIfsPass &pass, Block &rootBlock)
+      : pass(pass), rootBlock(rootBlock) {}
+  LogicalResult run();
+  void sinkOps();
+  void mergeIfs();
+};
+} // namespace
+
+void MergeIfsPass::runOnOperation() {
+  // Go through the regions recursively, from outer regions to nested regions,
+  // and try to move/sink/merge ops in each.
+  auto result = getOperation()->walk<WalkOrder::PreOrder>([&](Region *region) {
+    if (region->hasOneBlock() && mlir::mayHaveSSADominance(*region))
+      if (failed(Sinker(*this, region->front()).run()))
+        return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (result.wasInterrupted())
+    signalPassFailure();
+}
+
+/// Iteratively sink ops into block and move them closer to their uses, and
+/// merge adjacent `scf.if` operations.
+LogicalResult Sinker::run() {
+  LLVM_DEBUG(llvm::dbgs() << "Running on block in "
+                          << rootBlock.getParentOp()->getName() << "\n");
+  unsigned iteration = 0;
+  do {
+    if (++iteration > 10)
+      return rootBlock.getParentOp()->emitOpError()
```

Comment:
Under normal circumstances this should settled very quickly. Sinking operations and merging adjacent ifs can create additional optimization opportunities for each other. In the Rocket core this happens after 2-3 iterations, and when this hangs for too long it's usually an indication that there's a bug in the pass. But maybe the iteration count should rather be 100 or so.

Reply:
Actually, mabye this should just silently stop after hitting a maximum. You're right that this is an optimization and not essential to make the model work, so you probably don't want to abort the whole build because of some arbitrary limit.

Reply:
Or, thinking about this some more, maybe this shouldn't check an iteration count at all. If the pass works correctly it always converges. Otherwise it's a bug in the pass and you probably want this to just blow up.

Reply:
> you probably don't want to abort the whole build because of some arbitrary limit.

Agreed, it seems like a tricky thing to understand for a user who runs into it trying to get started with Arcilator. I think going without an iteration check makes sense personally.